src/Model/batchprocessing/batchprocessingMachineLearning/MachineLearningTrainingStage.py

```diff
@@ -0,0 +1,408 @@
+        random_forest_pred = rf_tree.predict(self.test_feature)
+
+        rmse_rf = np.sqrt(np.mean((self.test_label - random_forest_pred) ** 2))
+        score_rf = rf_tree.score(self.test_feature, self.test_label)
+
+        # MLP
+        grid_search_mlp.fit(self.train_feature, self.train_label)
+        mlp_model = grid_search_mlp.best_estimator_
+        mlp_pred = mlp_model.predict(self.test_feature)
+
+        rmse_mlp = np.sqrt(np.mean((self.test_label - mlp_pred) ** 2))
+        mlp_score = mlp_model.score(self.test_feature, self.test_label)
+
+        if rmse_mlp < rmse_rf:
+            self.score = mlp_score
+            self.accuracy['accuracy'] = f'{self.score}'
+            return mlp_model
+
+        self.score = score_rf
+        self.accuracy['accuracy'] = f'{self.score}'
+        return rf_tree
+
+    def regression_ml(self):
+        # RANDOM FOREST
+        forest_clas_reg = RandomForestRegressor(random_state=42)
+        forest_clas_reg.fit(self.train_feature, self.train_label)
+        random_forest_pred = forest_clas_reg.predict(self.test_feature)
+
+        rmse_rf = np.sqrt(np.mean((self.test_label - random_forest_pred) ** 2))
+        score_rf = forest_clas_reg.score(self.test_feature, self.test_label)
+
+        # MLP
+        mlp_reg = MLPRegressor(random_state=42, max_iter=5000)
+        mlp_reg.fit(self.train_feature, self.train_label)
+        mlp_pred = mlp_reg.predict(self.test_feature)
+
+        rmse_mlp = np.sqrt(np.mean((self.test_label - mlp_pred) ** 2))
+        mlp_score = mlp_reg.score(self.test_feature, self.test_label)
+
+        if rmse_mlp < rmse_rf:
+            self.score = mlp_score
+            self.accuracy['accuracy'] = f'{self.score}'
+            return mlp_reg
+
+        self.score = score_rf
+        self.accuracy['accuracy'] = f'{self.score}'
+        return forest_clas_reg
+
+    def save_ml_model(self, params_model_name, path, scaling):
+        # Create directory
+        list_ToSave = params_model_name["columns"]
+        list_ToSave.append(params_model_name["saveModel"])
+        dir_name = f'{path}{params_model_name["saveModel"]}'
+        try:
+            # Create target Directory
+            os.mkdir(dir_name)
+            print("Directory ", dir_name, " Created ")
+        except FileExistsError:
+            logging.debug("Directory ", dir_name, " already exists")
+
+            # Save Params for Clinical Data
+        with open(f'{dir_name}/'
+                  f'{params_model_name["saveModel"]}_params.txt',
+                  'w') as f:
+            print(list_ToSave, file=f)
+
+            # save in local directory
+            job_ml_file_ml = f'{dir_name}/' \
+                             f'{params_model_name["saveModel"]}_ml.pkl'
+            jobfile_scale = f'{dir_name}/' \
+                            f'{params_model_name["saveModel"]}_scaler.pkl'
+            joblib.dump(scaling, jobfile_scale)
+            # Save the model as a pickle in a file
```

Comment:
is the use of pickle format a requirement?
https://docs.python.org/3/library/pickle.html :

> Warning The pickle module is not secure. Only unpickle data you trust.
> It is possible to construct malicious pickle data which will execute arbitrary code during unpickling. Never unpickle data that could have come from an untrusted source, or that could have been tampered with.
> Consider signing data with [hmac](https://docs.python.org/3/library/hmac.html#module-hmac) if you need to ensure that it has not been tampered with.
> Safer serialization formats such as [json](https://docs.python.org/3/library/json.html#module-json) may be more appropriate if you are processing untrusted data. See [Comparison with json](https://docs.python.org/3/library/pickle.html#comparison-with-json).

I'm not recommending JSON.
Odds are this is going to be a big file and should be stored in a compact (binary) format.  

https://towardsdatascience.com/5-different-ways-to-save-your-machine-learning-model-b7996489d433
Unfortunately, it isn't clear if using joblib avoids the pickle module security issues.
https://scikit-learn.org/stable/model_persistence.html seems to suggest the same security issues, and indicates that version changes can leave your stored model unusable.
PMML looks like it might be more flexible and secure, although it's not going to be compact.

You might want to generate the HMAC and provide that through a separate channel if you intend to deliver the model to others, and provide a *means* for the user to provide the HMAC to test the .pkl before they try to load it.

Reply:
Yes, machine learning should be saved in the pkl format to be reused by scikit learn joblib

Joblib is an alternative tool to pickle that we can use to save [and load] our models.